tests/integration/test_charm.py

```diff
@@ -78,14 +83,24 @@ async def test_prometheus_grafana_integration(ops_test: OpsTest):
     ]
     log.info(f"Prometheus available at http://{prometheus_unit_ip}:9090")
 
-    r = requests.get(
-        f'http://{prometheus_unit_ip}:9090/api/v1/query?query=up{{juju_application="{APP_NAME}"}}'
-    )
-    response = json.loads(r.content.decode("utf-8"))
-    response_status = response["status"]
-    log.info(f"Response status is {response_status}")
-    assert response_status == "success"
+    try:
+        for attempt in Retrying(
+            (stop_after_attempt(5) | stop_after_delay(30)),
+            wait=wait_exponential(multiplier=1, min=1, max=10),
+            reraise=True,
+        ):
+            with attempt:
```

Comment:
Nothing wrong with this approach at all, but to me I found [this way](https://github.com/canonical/minio-operator/pull/56/files#diff-08291d7f4ad7b2caecacb6ae26af540ebbaede9fe42d5e8e1b39d2bdefe4c2d4R112) to make the test a little more readable (it just encapsulates the logic of the retry into a helper so you don't need to look at it unless you're interested).  Your choice whether you use this suggestion or not

Reply:
To be fair though, the other way has some odd `for: with:` syntax so it isn't perfect either

Reply:
I used your suggestion, that's indeed more readable. Thanks for providing an example